src/context.rs

```diff
@@ -76,6 +77,120 @@ impl Context {
+        insert: T,
+        insert2: T,
+    ) -> String {
+        self.stock_str(id)
+            .replacen("%1$s", insert.as_ref(), 1)
+            .replacen("%1$d", insert.as_ref(), 1)
+            .replacen("%2$s", insert2.as_ref(), 1)
+            .replacen("%2$d", insert2.as_ref(), 1)
+    }
+
+    /// Return some kind of stock message
+    ///
+    /// If the `id` is [StockId::MsgAddMember] or
+    /// [StockId::MsgDelMember] then `param1` is considered to be the
+    /// contact address and will be replaced by that contact's display
+    /// name.
+    ///
+    /// If `from_id` is not `0`, any trailing dot is removed from the
+    /// first stock string created so far.  If the `from_id` contact is
+    /// the user itself, i.e. `DC_CONTACT_ID_SELF` the string is used
+    /// itself as param to the [StockId::MsgActionByMe] stock string
+    /// resulting in a string like "Member Alice added by me." (for
+    /// [StockId::MsgAddMember] as `id`).  If the `from_id` contact
+    /// is any other user than the contact's display name is looked up and
+    /// used as the second parameter to [StockId::MsgActionByUser] with
+    /// again the original stock string being used as the first parameter,
+    /// resulting in a string like "Member Alice added by Bob.".
+    pub fn stock_system_msg<T: AsRef<str>>(
+        &self,
+        id: dc_stock::StockId,
+        param1: T,
+        param2: T,
+        from_id: u32,
+    ) -> String {
+        let insert1 = if id == StockId::MsgAddMember || id == StockId::MsgDelMember {
+            unsafe {
+                let param1_c = to_cstring(&param1);
+                let contact_id = dc_lookup_contact_id_by_addr(self, param1_c.as_ptr());
+                if contact_id != 0 as libc::c_uint {
+                    let contact = dc_get_contact(self, contact_id);
+                    let displayname = dc_contact_get_name_n_addr(contact);
+                    let ret = to_string(displayname);
+                    free(contact as *mut libc::c_void);
+                    free(displayname as *mut libc::c_void);
+                    ret
+                } else {
+                    param1.as_ref().to_string()
+                }
+            }
+        } else {
+            param1.as_ref().to_string()
+        };
+        let action = self.stock_str_repl_string2(id, insert1, param2.as_ref().to_string());
+        let action1 = action.trim_end_matches('.');
+        match from_id {
+            0 => action,
+            1 => self.stock_str_repl_string(StockId::MsgActionByMe, action1), // DC_CONTACT_ID_SELF
+            _ => unsafe {
+                let contact = dc_get_contact(self, from_id);
+                let displayname = dc_contact_get_display_name(contact);
+                let ret = self.stock_str_repl_string2(StockId::MsgActionByUser, action1, as_str(displayname));
+                free(contact as *mut libc::c_void);
+                free(displayname as *mut libc::c_void);
```

Comment:
a bit of a pity if new code uses c-data that needs to be freed etc.  Not sure though if trying to do a little refactoring to obtain a rusty "Contact" on which contact.display_name return a &str is in-inscope for this PR though.

Reply:
Yeah, this PR is about refactoring the stock strings to be safe.  Fixing this unsafe code would be in a PR (or multiple) which make the contact management safe.  Keeping PRs small and targetted...